Mathlib/RingTheory/Trace/Quotient.lean

```diff
@@ -0,0 +1,311 @@
+/-
+Copyright (c) 2024 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang, Riccardo Brasca
+-/
+
+import Mathlib.RingTheory.IntegralRestrict
+import Mathlib.RingTheory.DedekindDomain.Dvr
+
+/-!
+
+We gather results about the relations between the trace map on `B → A` and the trace map on
+quotients and localizations.
+
+## Main Results
+  `Algebra.trace_quotient_eq_of_isDedekindDomain` : The trace map on `B → A` coincides with the
+    trace map on `B⧸pB → A⧸p`.
+
+-/
+
+variable {R S : Type*} [CommRing R] [CommRing S] [Algebra R S] [LocalRing R]
+
+open LocalRing FiniteDimensional Submodule
+
+section LocalRing
+
+local notation "p" => maximalIdeal R
+local notation "pS" => Ideal.map (algebraMap R S) p
+
+variable [Module.Free R S] [Module.Finite R S]
+
+attribute [local instance] Ideal.Quotient.field
+
+theorem quotient_span_eq_top_iff_span_eq_top_of_localRing (s : Set S) :
+    span (R ⧸ p) ((Ideal.Quotient.mk (I := pS)) '' s) = ⊤ ↔
+    span R s = ⊤ := by
+  have H : (span (R ⧸ p) ((Ideal.Quotient.mk (I := pS)) '' s)).restrictScalars R =
+      (span R s).map (IsScalarTower.toAlgHom R S (S ⧸ pS)) := by
+    rw [map_span, ← restrictScalars_span R (R ⧸ p) Ideal.Quotient.mk_surjective]
+    rfl
+  constructor
+  · intro hs
+    rw [← top_le_iff]
+    apply le_of_le_smul_of_le_jacobson_bot
+    · exact Module.finite_def.mp ‹_›
+    · exact (jacobson_eq_maximalIdeal ⊥ bot_ne_top).ge
+    · rw [Ideal.smul_top_eq_map]
+      rintro x -
+      have : LinearMap.ker (IsScalarTower.toAlgHom R S (S ⧸ pS)) =
+          restrictScalars R pS := by
+        ext; simp [Ideal.Quotient.eq_zero_iff_mem]
+      rw [← this, ← comap_map_eq, mem_comap, ← H, hs]
+      trivial
+  · intro hs
+    rw [hs, Submodule.map_top] at H
+    change _ = LinearMap.range (Ideal.Quotient.mkₐ _ _) at H
+    rwa [LinearMap.range_eq_top.mpr (Ideal.Quotient.mkₐ_surjective _ _),
+      restrictScalars_eq_top_iff] at H
+
+theorem finrank_quotient_map_of_localRing :
```

Comment:
This lemma (and the previous one) also should have better places to go.

Reply:
Hmm, this is also a mess. Let me split some stuff before going back to this PR.

Reply:
Done